File: 2015/18/rust/src/main.rs
use std::fs;

fn gen_grid(path: &str) -> Vec<Vec<char>> {
    let data = fs::read_to_string(path)
        .expect("Error parsing input file")
        .trim()
        .lines()
        .map(|l| l.chars().collect())
        .collect();
    data
}

fn simulate_step(grid: &Vec<Vec<char>>, stuck: bool) -> (Vec<Vec<char>>, i32) {
    let rows = grid.len();
    let cols = grid[0].len();
    
    let mut simulated = grid.clone();
    let mut new_grid = grid.clone();

    // Set corner lights if stuck
    if stuck {
        for &(r, c) in &[(0, 0), (0, cols - 1), (rows - 1, 0), (rows - 1, cols - 1)] {
            new_grid[r][c] = '#';
            simulated[r][c] = '#';
        }
    }

    let neighbors = [
        (-1, -1), // top-left
        (-1, 0),  // top
        (-1, 1),  // top-right
        (0, -1),  // left
        (0, 1),   // right
        (1, -1),  // bottom-left
        (1, 0),   // bottom
        (1, 1),   // bottom-right
    ];

    for r in 0..rows {
        for c in 0..cols {
            // Skip stuck corners
            if stuck && [(0, 0), (0, cols - 1), (rows - 1, 0), (rows - 1, cols - 1)]
                .contains(&(r, c))
            {
                continue;
            }

            let n_lights = neighbors
                .iter()
                .filter(|&&(dr, dc)| {
                    let rr = r as i32 + dr;
                    let cc = c as i32 + dc;
                    rr >= 0 && rr < rows as i32 && cc >= 0 && cc < cols as i32
                        && simulated[rr as usize][cc as usize] == '#'
                })
                .count();

            if simulated[r][c] == '#' && !(n_lights == 2 || n_lights == 3) {
                new_grid[r][c] = '.';
            } else if simulated[r][c] == '.' && n_lights == 3 {
                new_grid[r][c] = '#';
            }
        }
    }

    let lights_on = new_grid
        .iter()
        .map(|r| r.iter().filter(|&&c| c == '#').count() as i32)
        .sum();

    (new_grid, lights_on)
}

fn simulate_n_steps(grid: &Vec<Vec<char>>, n_steps: i32, stuck: bool) -> i32 {
    let mut grid = grid.clone();
    let mut final_count = 0;
    
    for _ in 0..n_steps {
        let (new_grid, count) = simulate_step(&grid, stuck);
        grid = new_grid;
        final_count = count;
    }
    
    final_count
}

fn main() {
    let grid = gen_grid("../input.txt");
    let sep = "=".repeat(20);
    
    let total_1 = simulate_n_steps(&grid, 100, false);
    let total_2 = simulate_n_steps(&grid, 100, true);
    
    println!("{} Part 1 {}\nTotal lights on after 100 steps and not stuck: {}", sep, sep, total_1);
    println!("{} Part 2 {}\nTotal lights on after 100 steps and stuck: {}", sep, sep, total_2);
}

File: 2015/18/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func genGrid(path string) [][]rune {
	file, err := os.Open(path)
	if err != nil {
		panic("Error parsing input file")
	}
	defer file.Close()

	var data [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		data = append(data, []rune(scanner.Text()))
	}

	return data
}

func simulateStep(grid [][]rune, stuck bool) ([][]rune, int) {
	rows, cols := len(grid), len(grid[0])
	simulated := make([][]rune, rows)
	newGrid := make([][]rune, rows)

	for i := range grid {
		simulated[i] = make([]rune, cols)
		newGrid[i] = make([]rune, cols)
		copy(simulated[i], grid[i])
		copy(newGrid[i], grid[i])
	}

	if stuck {
		corners := [][2]int{{0, 0}, {0, cols - 1}, {rows - 1, 0}, {rows - 1, cols - 1}}
		for _, c := range corners {
			newGrid[c[0]][c[1]] = '#'
			simulated[c[0]][c[1]] = '#'
		}
	}

	neighbors := [][2]int{
		{-1, -1}, {-1, 0}, {-1, 1},
		{0, -1}, {0, 1},
		{1, -1}, {1, 0}, {1, 1},
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if stuck && ((r == 0 && c == 0) || (r == 0 && c == cols-1) || (r == rows-1 && c == 0) || (r == rows-1 && c == cols-1)) {
				continue
			}

			nLights := 0
			for _, n := range neighbors {
				rr, cc := r+n[0], c+n[1]
				if rr >= 0 && rr < rows && cc >= 0 && cc < cols && simulated[rr][cc] == '#' {
					nLights++
				}
			}

			if simulated[r][c] == '#' && !(nLights == 2 || nLights == 3) {
				newGrid[r][c] = '.'
			} else if simulated[r][c] == '.' && nLights == 3 {
				newGrid[r][c] = '#'
			}
		}
	}

	lightsOn := 0
	for _, row := range newGrid {
		for _, cell := range row {
			if cell == '#' {
				lightsOn++
			}
		}
	}

	return newGrid, lightsOn
}

func simulateNSteps(grid [][]rune, nSteps int, stuck bool) int {
	finalCount := 0
	for i := 0; i < nSteps; i++ {
		grid, finalCount = simulateStep(grid, stuck)
	}
	return finalCount
}

func main() {
	grid := genGrid("../input.txt")
	sep := strings.Repeat("=", 20)
	total1 := simulateNSteps(grid, 100, false)
	total2 := simulateNSteps(grid, 100, true)

	fmt.Printf("%s Part 1 %s\nTotal lights on after 100 steps and not stuck: %d\n", sep, sep, total1)
	fmt.Printf("%s Part 2 %s\nTotal lights on after 100 steps and stuck: %d\n", sep, sep, total2)
}
